_posts/2015-10-12-creating-bem-stylesheets-for-react-components.markdown

```diff
@@ -0,0 +1,279 @@
+categories:   frontend
+---
+
+There are plenty of options to choose from when it comes to frontend methodologies.  As I've learned while working with Ember and React.js, web interfaces are composed of powerful components.  When thinking about how to approach the organization of a project, I feel the CSS for each component needs to be just as concise and clear as the Javascript that creates it.
+
+Enter BEM: it stands for "Block", "Element", and "Modifier".  This is no new approach, but it's really begun to grow on me.  I've started to create BEM-style "component" stylesheets here at WeWork, and my main goal is to create "blocks" that are independent of each other, wholey reuseable. Remember, a block may be developed as a singular unit, but it can and **should** be able to appear on a page more than once.
+
+A **block** is *independent* and can be utilized on its own or can contain other blocks (ie. `.location-info {}`.) These can be thought of as parents. The children of blocks are their **elements**, written with two underscores (ie. `.location-info__content {}`.)  **Modifiers** specify the type of block that we're dealing with, so that we can style that component without affecting others (ie. `.location-info--with-background`.)
+
+The markup might look like this:
+
+
+```html
+<div class="location-info location-info--with-background">
+    <div class="location-info__content">
+        <div class="location-info__image">
+            <img src="..." />
+        </div>
+        <div class="location-info__text">
+            <p>Welcome to Soho West!</p>
+        </div>
+    </div>
+</div>
+```
+
+
+
+
+A few things to remember:
+
+- HTML Elements must not be used in CSS selectors. We want our classes to be context-free. (ie. In the example above, I should not be applying a style to `.location-info__image img`.)
+- ID-based selectors must not be used. We are developing components that are non-unique! There could be multiple blocks/elements on one page.
+- Avoid making parent elements when the child can exist by itself.
+- You do not need to use BEM for absolutely everything. Just because there is an element in a block, it does not mean it is a BEM element.  Some things just won't fall under the category of BEM. For example, take the class: `.text-center { text-align: center; }`. This is a standalone class, not an element. No need to BEM it up!
+
+
+## BEM in Action
+
+The BEM methodology has been very useful within our React components. Seeing as we're already creating objects that are wholey reuseable across the site, it only seems natural to make sure that our CSS is as concise.
+
+I was recently creating a new React component for our "Book a Tour" modal, called LocationCmInfo ("Cm" standing for Community Manager, the person giving the tour.)
+
+
+```js
+import React, {Component, PropTypes} from 'react';
+
+class LocationCmInfo extends Component {
```

Comment:
Let's chat about cleaning this up tomorrow! I feel like for the purposes of this article the reader shouldn't have to know about ES6 or jsx for that matter.

Let's just assume that we're building a component with just html, and also include an image of the finished component as a frame of reference.